### PR TITLE
Initial `Prepare` and `Finish` steps implementation; VagrantProvision and `run.sh` enhancements

### PR DESCRIPTION
**Pull request by pvalena**
Updated: 2019-11-06T16:43+00:00

### CODE REVIEW

**Comment by pvalena** [comment]
Updated with #32 and #31.

**Comment by pvalena** [comment]
Feel free to merge this one....
Joined:  #32 and  #37

**Comment by pvalena** [comment]
Resolved.

**Comment by psss** [comment]
Tests are failing:

    cannot import name 'join' from 'shlex'

File: tmt/steps/provision/localhost.py [hunk 1]
```diff
@@ -1,8 +1,8 @@
 from click import echo
 
 from tmt.steps.provision.base import ProvisionBase
-
+from shlex import join
```
Comment (thrix) [inline tmt/steps/provision/localhost.py]: we need to find something that works here for python 3.6 and 3.7. This was added in 3.8.

I currently just use as a workaround:

```
from os.path import join
```

Comment (psss) [reply tmt/steps/provision/localhost.py]: The `os.path.join` would not work here as it joins with slashes. We could use `shlex.quote` or even get rid of the join.

Comment (pvalena) [reply tmt/steps/provision/localhost.py]: I must've just forgotten to remove this. I've reimplemented it myself. And no it cannot be replaced with `os.path.join` AFAIK.

Comment (pvalena) [reply tmt/steps/provision/localhost.py]: Here: https://github.com/psss/tmt/pull/30/files#diff-253b78529284ad6a69cc8e0bf8d0df02R55-R61

Comment (pvalena) [reply tmt/steps/provision/localhost.py]: But that was already merged, wasn't it?

**Comment by psss** [comment]
Thanks Pavel. I've reviewed and tested the patch. Looks good and works. Have some improvement ideas but let's work them later. Squashed and after a couple of changes in dff8982 merged into master.

**Comment by psss** [comment]
Example output of `tmt run --verbose --all provision --how=libvirt` looks nice:

    /var/tmp/tmt/run-756
    Found 3 plans.
    
    /plans/basic
    summary: Essential command line features
        discover
            how: fmf
            repository: https://github.com/psss/tmt
            revision: devel
            filter: tier: 0,1
            tests: 2 tests selected
                /one/tests/docs
                /one/tests/ls
        provision
        prepare
            Prepare: ansible = 'plans/packages.yml
        execute
            how: beakerlib
            overview: FF
            result: 0 tests passed, 2 tests failed
        report
        finish
    
    /plans/helps
    summary: Check help messages
        discover
            how: shell
            directory: /home/psss/git/tmt
            tests: 4 tests selected
                /help/main
                /help/test
                /help/plan
                /help/smoke
        provision
        prepare
            Prepare: shell = 'dnf copr enable -y psss/tmt && dnf install -y tmt
        execute
            how: shell
            overview: ....
            result: 4 tests passed, 0 tests failed
        report
        finish
    
    /plans/smoke
    summary: Just a basic smoke test
        discover
            how: shell
            tests: 0 tests selected
        provision
        prepare
        execute
            how: shell
            overview: E
            result: 0 tests passed, 0 tests failed
        report
        finish

Shows both supported prepares (shell & ansible) and passed, failed and errored tests.

**Comment by pvalena** [comment]
great!

**Pull request closed without merging**